Added SVM Classifier

John42506176Linux opened the pull request:
SVM classifier has been fully added with probability for f1 score , precision and recall metrics.
<img width="558" alt="SVMtestoutput" src="https://user-images.githubusercontent.com/20214982/97105667-1dbf1e00-1693-11eb-9c68-2261dd7fdea3.PNG">
<img width="312" alt="SVMoutput" src="https://user-images.githubusercontent.com/20214982/97105817-3d0a7b00-1694-11eb-91ab-706905ea855b.PNG">

Closes #99

Andrey1994 commented on python-package/brainflow/ml_model.py:
```diff
@@ -27,7 +27,8 @@ class BrainFlowClassifiers (enum.Enum):
 
     REGRESSION = 0 #:
-    KNN = 1 #:
+    KNN = 1  #:
```
extra space here

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+#include <algorithm>
+#include <cmath>
+#include <stdlib.h>
+#include <iostream>
```
dont need iostream

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+#include <cmath>
+#include <stdlib.h>
+#include <iostream>
+#include "brainflow_constants.h"
```
lets add a new line before local includes

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+    char path[1024];
+    bool res = get_dll_path (path);
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
```
this string looks like clang-format doesnt work

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+    char path[1024];
+    bool res = get_dll_path (path);
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
+    std::strcpy (fullPath,path);
```
as well as here and below, I belive clang-format adds a space after comma

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+{
+    if ((data_len < 5) || (data == NULL) || (output == NULL))
+    {
+        safe_logger (spdlog::level::err, "Classifier has already been prepared.");
```
log message is wrong

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+
+int ConcentrationSVMClassifier::predict (double *data, int data_len, double *output)
+{
+    if ((data_len < 5) || (data == NULL) || (output == NULL))
```
does it work if data_len is 5 instead 10?

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+        return (int)BrainFlowExitCodes::INVALID_BUFFER_SIZE_ERROR;
+    }
+    struct svm_node *x;
+    x =  (struct svm_node *) malloc(data_len*sizeof(struct svm_node));
```
looks like memory leak, I dont see free call in the end. You can try to run it with valgrind to test

Andrey1994 replied:
and clang-format is definitely off on your PC

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+
+int ConcentrationSVMClassifier::release ()
+{
+    delete model;
```
add a check that its not NULL or that classifier is prepared

Andrey1994 commented on src/ml/concentration_svm_classifier.cpp:
```diff
@@ -0,0 +1,49 @@
+    char * fullPath = new char [std::strlen(path)+std::strlen("brainflow_svm.model")+1];
+    std::strcpy (fullPath,path);
+    std::strcat (fullPath ,"brainflow_svm.model");
+    model = svm_load_model (fullPath);
```
we use _ in naming instead camel case

Andrey1994 commented on tests/cpp/ml_demo/src/eeg_metrics.cpp:
```diff
@@ -11,6 +11,7 @@
 #include "board_shim.h"
 #include "data_filter.h"
 #include "ml_model.h"
+#include  <vector>
```
can we move SVM to another test\file? This example is used in docs on public page as code sample. I want to keep it short and simple for demo purposes

Andrey1994 commented on tests/python/eeg_metrics.py:
```diff
@@ -38,9 +39,14 @@ def main ():
     params.file = args.file
-
-    board = BoardShim (args.board_id, params)
-    sampling_rate = BoardShim.get_sampling_rate (args.board_id)
```
and the same here, lets move to another file. I really want to keep code samples from docs less than one screen of code

Andrey1994 commented:
Also, it will not work on android and its ok but we should add smth like 
```
#ifdef Android
return UNSUPPORTED....
```

John42506176Linux commented:
Also here's a comment on how I addressed the less simple things in the PR

-I created a full version for eeg_metrics for the CPP and python file, the eeg_metrics only has added the master_board_id thing we talked about.

-I added everything needed for Valgrind so it should be fine now, the only thing I saw was a slight issue for the KNN classifier regarding the safe logger.

- It will work if data is less than 10 but considering all the other classifiers use it, I thought it'd be better to keep consistency.

-All of the svm_classifier have added the if_def android.

John42506176Linux commented:
Just fixed some minor log issues valgrind runs perfectly now.
<img width="508" alt="ValgrindTest" src="https://user-images.githubusercontent.com/20214982/97211224-18e29300-1795-11eb-891b-e22f467f0adc.PNG">

Andrey1994 commented:
lets revert commit https://github.com/brainflow-dev/brainflow/pull/112/commits/e9a42c2c7e2c7556d6c77ad7914caebf623081d3 and make it like here https://github.com/John42506176Linux/brainflow/pull/2 probably in all bindings

Andrey1994 commented:
CI is down due to an error in java test

Andrey1994 commented:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project brainflow: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/brainflow-dev/brainflow/java-package/brainflow/src/main/java/brainflow/examples/EEGMetrics.java:[54,30] variable model_params is already defined in method main(java.lang.String[])
[ERROR] /home/travis/build/brainflow-dev/brainflow/java-package/brainflow/src/main/java/brainflow/examples/EEGMetrics.java:[56,17] variable concentration is already defined in method main(java.lang.String[])
```
